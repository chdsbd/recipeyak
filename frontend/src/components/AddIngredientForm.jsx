import React from 'react'

import { ButtonPrimary } from './Buttons'

const AddIngredientForm = ({
  handleAddIngredient,
  cancelAddIngredient,
  handleInputChange,
  quantity,
  name,
  description,
  optional,
  loading,
  error,
}) =>
  <form onSubmit={ async e => {
    e.preventDefault()
    await handleAddIngredient()
    document.querySelector('#firstinput').focus()
  }
  }>
  <div className="add-ingredient-grid mb-2 mt-2">

    <div>
      <input
        id="firstinput"
        onChange={ handleInputChange }
        onFocus={ e => e.target.select() }
        autoFocus
        value={ quantity }
        className={ 'my-input' + (error ? ' is-danger' : '') }
        type="text"
        placeholder="3 lbs"
        name="quantity"/>
    </div>

    <div>
      <input
        onChange={ handleInputChange }
        onFocus={ e => e.target.select() }
        value={ name }
        className={ 'my-input' + (error ? ' is-danger' : '') }
        type="text"
        placeholder="tomato"
        name="name"/>
    </div>

    <div className="grid-entire-row">
      <input
        onChange={ handleInputChange }
        onFocus={ e => e.target.select() }
        value={ description }
        className={ 'my-input' + (error ? ' is-danger' : '') }
        type="text"
        placeholder="diced at 3cm"
        name="description"/>
      { error
          ? <p className="fs-4 c-danger">A recipe needs at least one ingredient</p>
          : null
      }
    </div>
  </div>

  <label className="d-flex align-items-center cursor-pointer mb-2">
  <input
    onChange={ handleInputChange }
    checked={ optional }
    name="optional"
    type='checkbox'
    className="mr-2"/>
    Optional
  </label>

  <div className="field is-grouped">
    <p className="control">
      <ButtonPrimary
        disabled={ quantity === '' && name === '' }
        className="is-small"
        type="submit"
        name="add ingredient"
        loading={ loading }>
        Add
      </ButtonPrimary>
    </p>
    <p className="control">
      <input
        onClick={ cancelAddIngredient }
        className="my-button is-small"
        type="button"
        name="cancel add ingredient"
        value="cancel"/>
    </p>
  </div>
</form>

export default AddIngredientForm
