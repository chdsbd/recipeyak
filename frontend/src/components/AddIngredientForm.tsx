import * as React from 'react'

import { ButtonPrimary } from './Buttons'

const AddIngredientForm = ({
  handleAddIngredient,
  cancelAddIngredient,
  handleInputChange,
  quantity,
  name,
  description,
  loading,
  error
}) =>
  <form onSubmit={ async e => {
    e.preventDefault()
    await handleAddIngredient()
    document.querySelector('#firstinput').focus()
  }
  }>
  <div className="add-ingredient-grid mb-2 mt-2">

    <div>
      <input
        id="firstinput"
        onChange={ handleInputChange }
        onFocus={ e => e.target.select() }
        value={ quantity }
        className={ 'my-input' + (error ? ' is-danger' : '') }
        type="text"
        placeholder="3 lbs"
        name="quantity"/>
    </div>

    <div>
      <input
        onChange={ handleInputChange }
        onFocus={ e => e.target.select() }
        value={ name }
        className={ 'my-input' + (error ? ' is-danger' : '') }
        type="text"
        placeholder="tomato"
        name="name"/>
    </div>

    <div className="grid-entire-row">
      <input
        onChange={ handleInputChange }
        onFocus={ e => e.target.select() }
        value={ description }
        className={ 'my-input' + (error ? ' is-danger' : '') }
        type="text"
        placeholder="diced at 3cm"
        name="description"/>
      { error
          ? <p className="fs-4 c-danger">A recipe needs at least one ingredient</p>
          : null
      }
    </div>
  </div>

  <div className="field is-grouped">
    <p className="control">
      <ButtonPrimary
        disabled={ quantity === '' && name === '' }
        type="submit"
        name="add ingredient"
        loading={ loading }>
        Add
      </ButtonPrimary>
    </p>
    <p className="control">
      { quantity !== '' || name !== ''
        ? <input
            disabled={ quantity === '' && name === '' }
            onClick={ cancelAddIngredient }
            className="my-button"
            type="button"
            name="cancel add ingredient"
            value="âœ•"/>
        : null
      }
    </p>
  </div>
</form>

export default AddIngredientForm
