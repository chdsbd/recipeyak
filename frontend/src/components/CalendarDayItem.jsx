import React from "react"
import PropTypes from "prop-types"
import { Link } from "react-router-dom"
import { DragSource } from "react-dnd"

import { beforeCurrentDay } from "../date"

import { recipeURL } from "../urls"

import * as DragDrop from "../dragDrop"

const COUNT_THRESHOLD = 1

const source = {
  beginDrag(props) {
    return {
      recipeID: props.recipeID,
      count: props.count,
      id: props.id
    }
  },
  canDrag({ date }) {
    return !beforeCurrentDay(date)
  },
  endDrag(props, monitor) {
    // when dragged onto something that isn't a target, we remove it
    if (!monitor.didDrop()) {
      props.remove()
    }
  }
}

function collect(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging()
  }
}

@DragSource(DragDrop.RECIPE, source, collect)
export default class CalendarItem extends React.Component {
  state = {
    count: this.props.count,
    hover: false
  }

  static propTypes = {
    id: PropTypes.number.isRequired,
    recipeID: PropTypes.number.isRequired,
    recipeName: PropTypes.string.isRequired,
    count: PropTypes.number.isRequired,
    remove: PropTypes.func.isRequired,
    updateCount: PropTypes.func.isRequired,
    refetchShoppingList: PropTypes.func.isRequired,
    date: PropTypes.object.isRequired
  }

  componentWillMount() {
    this.setState({ count: this.props.count })
    document.addEventListener("keypress", this.handleKeyPress)
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ count: nextProps.count })
  }

  componentWillUnmount() {
    document.removeEventListener("keypress", this.handleKeyPress)
  }

  handleKeyPress = e => {
    if (!this.state.hover) return

    if (beforeCurrentDay(this.props.date)) return

    if (e.key === "#") {
      this.props.remove()
    }
    if (e.key === "A") {
      this.updateCount(this.state.count + 1)
    }
    if (e.key === "X") {
      this.updateCount(this.state.count - 1)
    }
  }

  updateCount = count => {
    const oldCount = this.state.count
    if (beforeCurrentDay(this.props.date)) return
    this.setState({ count })
    this.props
      .updateCount(count)
      .then(() => this.props.refetchShoppingList())
      .catch(() => this.setState({ count: oldCount }))
  }

  handleChange = e => {
    const count = e.target.value
    this.updateCount(count)
  }

  render() {
    const { connectDragSource, isDragging } = this.props
    return connectDragSource(
      <div
        className="d-flex align-items-center cursor-pointer justify-space-between mb-1"
        onMouseEnter={() => this.setState({ hover: true })}
        onMouseLeave={() => this.setState({ hover: false })}
        style={{
          visibility: isDragging ? "hidden" : "visible"
        }}>
        <Link
          to={recipeURL(this.props.recipeID, this.props.recipeName)}
          className="break-word fs-3"
          style={{
            lineHeight: 1.1
          }}>
          {this.props.recipeName}
        </Link>
        {this.state.count > COUNT_THRESHOLD ? (
          <div className="d-flex">
            <input
              className="fs-3 my-input text-right w-2rem"
              name="calendar-item-count"
              onChange={this.handleChange}
              value={this.state.count}
            />
          </div>
        ) : null}
      </div>
    )
  }
}
