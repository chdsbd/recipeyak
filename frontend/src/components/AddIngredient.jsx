import React from 'react'
import AddIngredientForm from './AddIngredientForm'

class AddIngredient extends React.Component {
  constructor (props) {
    super(props)

    this.emptyState = {
      addingIngredient: false,
      quantity: '',
      unit: '',
      name: '',
      description: ''
    }

    this.state = this.emptyState
  }

  handleInputChange = e => {
    this.setState({ [e.target.name]: e.target.value })
  }

  handleAddIngredient = e => {
    e.preventDefault()
    this.addIngredient(this.props.id, this.state.ingredient)
  }

  addingIngredient = () => {
    this.setState({ addingIngredient: true })
  }

  cancelAddIngredient = () => {
    this.setState(this.emptyState)
  }

  clearInputs = () => {
    this.setState({ quantity: '', unit: '', name: '', description: '' })
  }

  addIngredient = async (id, ingredient) => {
    // TODO: pass this prop in from Recipe or via store
    const { quantity, unit, name, description } = this.state
    await this.props.addIngredient(id, { quantity, unit, name, description })
    this.clearInputs()
  }

  render () {
    const { state } = this

    const units = [
      'large',
      'small',
      'medium',
      'tsp',
      'tbsp',
      'cup',
      'lb',
      'kg',
      'mg',
      'gram',
      'liter',
      'ml',
      'gal.',
      'oz',
      'fl oz',
      'pint',
      'quart',
      'pinch',
      'inch',
      'cm',
      'm',
      'mm'
    ]

    return (
      state.addingIngredient
      ? <AddIngredientForm
          handleAddIngredient={this.handleAddIngredient}
          cancelAddIngredient={this.cancelAddIngredient}
          handleInputChange={this.handleInputChange}
          units={units}
          quantity={this.state.quantity}
          unit={this.state.unit}
          name={this.state.name}
          description={this.state.description}
        />
      : <p className="flex-center">
        <button
          onClick={ this.addingIngredient }
          className="button is-link">
          Add another
        </button>
      </p>
    )
  }
}

export default AddIngredient
