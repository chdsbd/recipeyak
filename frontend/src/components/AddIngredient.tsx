import React from "react"
import PropTypes from "prop-types"
import AddIngredientForm from "./AddIngredientForm"

export interface IIngredientBasic {
  quantity: string
  name: string
  description: string
}

interface IAddIngredientProps {
  readonly onCancel: () => void
  readonly id: number
  readonly addIngredient: (
    id: number,
    { quantity, name, description }: IIngredientBasic
  ) => void
  readonly loading: boolean
  readonly autoFocus: boolean
}

interface IAddIngredientState {
  readonly quantity: string
  readonly name: string
  readonly description: string
  readonly optional: boolean
  readonly addingIngredient: boolean
}

export type AddIngredientFields = Exclude<
  keyof IAddIngredientState,
  "addingIngredient"
>

export default class AddIngredient extends React.Component<
  IAddIngredientProps,
  IAddIngredientState
> {
  static propTypes = {
    id: PropTypes.number.isRequired,
    addIngredient: PropTypes.func.isRequired,
    loading: PropTypes.bool.isRequired,
    autoFocus: PropTypes.bool,
    onCancel: PropTypes.func.isRequired
  }

  static defaultProps = {
    loading: false
  }

  emptyState: IAddIngredientState = {
    quantity: "",
    name: "",
    description: "",
    optional: false,
    addingIngredient: false
  }

  state: IAddIngredientState = this.emptyState

  handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.setState(({
      [e.target.name]: e.target.value
    } as unknown) as IAddIngredientState)
  }

  addingIngredient = () => {
    this.setState({ addingIngredient: true })
  }

  cancelAddIngredient = () => {
    this.props.onCancel()
    this.setState(this.emptyState)
  }

  clearInputs = () => {
    this.setState(this.emptyState)
  }

  render() {
    const { clearInputs, cancelAddIngredient, handleInputChange } = this
    const { id, addIngredient, loading, autoFocus } = this.props
    const { quantity, name, description, optional } = this.state

    return (
      <AddIngredientForm
        handleAddIngredient={async () => {
          await addIngredient(id, { quantity, name, description })
          clearInputs()
        }}
        loading={loading}
        cancelAddIngredient={cancelAddIngredient}
        handleInputChange={handleInputChange}
        quantity={quantity}
        name={name}
        description={description}
        optional={optional}
        autoFocus={autoFocus}
      />
    )
  }
}
