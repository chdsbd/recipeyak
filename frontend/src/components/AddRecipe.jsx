import React from 'react'
import { Helmet } from 'react-helmet'

import ListItem from './ListItem'
import AddIngredientForm from './AddIngredientForm'
import AddStepForm from './AddStepForm'
import Ingredient from './Ingredient'
import { ButtonPrimary } from './Buttons'

const unfinishedIngredient = ({ quantity = '', name = '' }) =>
  quantity === '' || name === ''

class AddRecipe extends React.Component {
  state = {
    ingredient: {
      quantity: '',
      name: '',
      description: ''
    },
    step: ''
  }

  static defaultProps = {
    error: {
      errorWithName: false,
      errorWithIngredients: false,
      errorWithSteps: false
    },
    loading: false,
    name: '',
    author: '',
    source: '',
    time: '',
    servings: '',
    ingredients: [],
    steps: []
  }

  componentWillMount = () =>
    this.props.clearErrors()

  handleInputChange = e =>
    this.setState({ [e.target.name]: e.target.value })

  handleSubmit = event => {
    event.preventDefault()
    const recipe = {
      name: this.props.name,
      author: this.props.author,
      source: this.props.source,
      time: this.props.time,
      servings: this.props.servings,
      ingredients: this.props.ingredients,
      steps: this.props.steps
    }
    this.props.addRecipe(recipe)
  }

  addIngredient = () => {
    if (unfinishedIngredient(this.state.ingredient)) return
    this.props.addIngredient(this.state.ingredient)
    this.setState({ ingredient: {} })
  }

  handleIngredientChange = e => {
    e.persist()
    this.setState(prevState => ({
      ingredient: {
        ...prevState.ingredient,
        [e.target.name]: e.target.value
      }
    }))
  }

  cancelAddIngredient = () =>
    this.setState({ ingredient: {} })

  addStep = () => {
    this.props.addStep({ text: this.state.step })
    this.setState({ step: '' })
  }

  cancelAddStep = () =>
    this.setState({ step: '' })

  render () {
    const {
      ingredient,
      step
    } = this.state

    const {
      addStep,
      handleInputChange,
      cancelAddStep,
      handleSubmit,
      addIngredient,
      cancelAddIngredient,
      handleIngredientChange
    } = this

    const {
      errorWithName,
      errorWithIngredients,
      errorWithSteps
    } = this.props.error

    const {
      quantity = '',
      name = '',
      description = ''
    } = ingredient

    return (
      <div className="d-grid grid-gap-1rem">
        <Helmet title='Add Recipe'/>
        <div>
          <input
            autoFocus
            onChange={ this.props.setName }
            value={ this.props.name }
            className={ 'my-input fs-2rem' + (errorWithName ? ' is-danger' : '') }
            type="text" placeholder="new recipe title" name="name"/>
          { errorWithName
              ? <p className="fs-4 c-danger">A recipe needs a name</p>
              : null
          }
        </div>

        <div className="d-grid  meta-data-grid">
          <label className="d-flex align-center">By
            <input
              onChange={ this.props.setAuthor }
              value={ this.props.author }
              className="my-input ml-2"
              type="text"
              placeholder="Author"
              name="author"/>
          </label>
          <label className="d-flex align-center">from
            <input
              onChange={ this.props.setSource }
              value={ this.props.source }
              className="my-input ml-2"
              type="text"
              placeholder="http://example.com/dumpling-soup"
              name="source"/>
          </label>
          <label className="d-flex align-center">creating
            <input
              onChange={ this.props.setServings }
              value={ this.props.servings }
              className="my-input ml-2"
              type="text"
              placeholder="4 to 6 servings"
              name="servings"/>
          </label>
          <label className="d-flex align-center">in
            <input
              onChange={ this.props.setTime }
              value={ this.props.time }
              className="my-input ml-2"
              type="text"
              placeholder="1 hour"
              name="time"/>
          </label>
        </div>

          <section className="ingredients-preparation-grid">
            <div>
              <h2 className="title">Ingredients</h2>
              <ul>
                {
                  this.props.ingredients.map((ingredient, i) =>
                    <Ingredient
                      key={ ingredient.name + i }
                      index={ i }
                      id={ i }
                      update={ this.props.updateIngredient }
                      remove={ this.props.removeIngredient }
                      quantity={ ingredient.quantity }
                      name={ ingredient.name }
                      description={ ingredient.description }
                    />
                  )
                }
              </ul>

              <AddIngredientForm
                handleAddIngredient={ addIngredient }
                cancelAddIngredient={ cancelAddIngredient }
                handleInputChange={ handleIngredientChange }
                quantity={ quantity }
                name={ name }
                description={ description }
                error={ errorWithIngredients }
              />
            </div>

              <div>
                <h2 className="title is-3">Preparation</h2>
                <ul>
                  {
                    this.props.steps.map((step, i) =>
                      <div key={step.text + i}>
                        <label className="label">Step { i + 1}</label>
                        <ListItem
                          id={i}
                          text={step.text}
                          update={ this.props.updateStep }
                          delete={ this.props.removeStep }
                        />
                      </div>
                    )
                  }
                </ul>
                <AddStepForm
                  handleInputChange={ handleInputChange }
                  addStep={ addStep }
                  cancelAddStep={ cancelAddStep }
                  stepNumber={ this.props.steps.length + 1 }
                  text={ step }
                  error={ errorWithSteps }
                />
            </div>
          </section>
        <div className="d-flex justify-space-between align-items-center">
          <button
            className='my-button'
            onClick={ this.props.clearForm }
            name="create recipe">
            Clear
          </button>

          <div className="d-flex justify-space-between">

          <label className="d-flex align-center">for
            <div className="select ml-2 is-small">
            <select >
              <option value="">Personal</option>
              <option value="">Team: Recipe Yak</option>
              <option value="">Team: Red Team</option>
              <option value="">Team: Blue Team</option>
            </select>
          </div>
          </label>

          <ButtonPrimary
            className="ml-2"
            type="submit"
            onClick={ handleSubmit }
            name="create recipe"
            loading={ this.props.loading }>
            Create Recipe
          </ButtonPrimary>
        </div>
        </div>
      </div>
    )
  }
}

export default AddRecipe
