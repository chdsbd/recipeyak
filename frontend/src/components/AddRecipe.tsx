import React from "react"
import { Helmet } from "@/components/Helmet"

import ListItem from "@/components/ListItem"
import AddIngredientForm from "@/components/AddIngredientForm"
import AddStepForm from "@/components/AddStepForm"
import Ingredient from "@/components/Ingredient"
import { ButtonPrimary } from "@/components/Buttons"
import {
  IStepBasic,
  IRecipeBasic,
  IIngredientBasic,
  IAddRecipeError
} from "@/store/reducers/recipes"
import { ITeam } from "@/store/reducers/teams"

const unfinishedIngredient = ({ quantity = "", name = "" }) =>
  quantity === "" || name === ""

interface IAddRecipeProps {
  readonly clearErrors: () => void
  readonly addStep: (step: IStepBasic) => void
  readonly clearForm: () => void
  readonly setTime: (e: React.ChangeEvent<HTMLInputElement>) => void
  readonly setServings: (e: React.ChangeEvent<HTMLInputElement>) => void
  readonly setSource: (e: React.ChangeEvent<HTMLInputElement>) => void
  readonly setAuthor: (e: React.ChangeEvent<HTMLInputElement>) => void
  readonly setName: (e: React.ChangeEvent<HTMLInputElement>) => void
  readonly removeStep: (index: number) => void
  readonly updateStep: (_recipeID: number, i: number, step: IStepBasic) => void
  readonly fetchData: () => void
  readonly addRecipe: (recipe: IRecipeBasic) => void
  readonly removeIngredient: (index: number) => void
  readonly updateIngredient: (
    index: number,
    ingredient: IIngredientBasic
  ) => void
  readonly addIngredient: (ingredient: IIngredientBasic) => void
  readonly error: IAddRecipeError
  readonly loading: boolean
  readonly name: string
  readonly author: string
  readonly source: string
  readonly time: string
  readonly servings: string
  readonly ingredients: IIngredientBasic[]
  readonly steps: IStepBasic[]
  readonly loadingTeams: boolean
  readonly teams: ITeam[]
  readonly setTeamID: (x: number | null) => void
  readonly teamID: ITeam["id"] | null
}

interface IAddRecipeState {
  readonly ingredient: {
    readonly quantity: string
    readonly name: string
    readonly description: string
    readonly optional: boolean
    readonly [key: string]: string | boolean
  }
  readonly step: string
}

const emptyIngredient = {
  quantity: "",
  name: "",
  description: "",
  optional: false
}

export default class AddRecipe extends React.Component<
  IAddRecipeProps,
  IAddRecipeState
> {
  state: IAddRecipeState = {
    ingredient: emptyIngredient,
    step: ""
  }

  componentDidMount() {
    this.props.clearErrors()
    this.props.fetchData()
  }

  handleInputChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    this.setState(({
      [e.target.name]: e.target.value
    } as unknown) as IAddRecipeState)
  }

  handleSubmit = (event: React.FormEvent) => {
    event.preventDefault()
    this.props.addRecipe({
      name: this.props.name,
      author: this.props.author,
      source: this.props.source,
      time: this.props.time,
      servings: this.props.servings,
      ingredients: this.props.ingredients,
      steps: this.props.steps,
      team: this.props.teamID || undefined
    })
  }

  addIngredient = () => {
    if (unfinishedIngredient(this.state.ingredient)) {
      return Promise.resolve()
    }
    this.props.addIngredient(this.state.ingredient)
    this.setState({ ingredient: emptyIngredient })
  }

  handleIngredientChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.persist()
    if (e.target.type === "checkbox") {
      this.setState(prev => ({
        ingredient: {
          ...prev.ingredient,
          [e.target.name]: !prev.ingredient[e.target.name]
        }
      }))
      return
    }
    this.setState(prevState => ({
      ingredient: {
        ...prevState.ingredient,
        [e.target.name]: e.target.value
      }
    }))
  }

  cancelAddIngredient = () => this.setState({ ingredient: emptyIngredient })

  addStep = () => {
    this.props.addStep({ text: this.state.step })
    this.setState({ step: "" })
  }

  cancelAddStep = () => this.setState({ step: "" })

  handleTeamChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    if (e.target.value === "personal") {
      return this.props.setTeamID(null)
    }
    const id = parseInt(e.target.value, 10)
    this.props.setTeamID(id)
  }

  render() {
    const { ingredient, step } = this.state

    const {
      addStep,
      handleInputChange,
      cancelAddStep,
      handleSubmit,
      addIngredient,
      cancelAddIngredient,
      handleIngredientChange
    } = this

    const {
      errorWithName,
      errorWithIngredients,
      errorWithSteps
    } = this.props.error

    const {
      quantity = "",
      name = "",
      description = "",
      optional = false
    } = ingredient

    return (
      <div className="d-grid grid-gap-1rem">
        <Helmet title="Add Recipe" />
        <div>
          <input
            autoFocus
            onChange={this.props.setName}
            value={this.props.name}
            className={"my-input fs-2rem" + (errorWithName ? " is-danger" : "")}
            type="text"
            placeholder="new recipe title"
            name="name"
          />
          {errorWithName ? (
            <p className="fs-4 c-danger">A recipe needs a name</p>
          ) : null}
        </div>

        <div className="d-grid  meta-data-grid">
          <label className="d-flex align-center">
            By
            <input
              onChange={this.props.setAuthor}
              value={this.props.author}
              className="my-input ml-2"
              type="text"
              placeholder="Author"
              name="author"
            />
          </label>
          <label className="d-flex align-center">
            from
            <input
              onChange={this.props.setSource}
              value={this.props.source}
              className="my-input ml-2"
              type="text"
              placeholder="http://example.com/dumpling-soup"
              name="source"
            />
          </label>
          <label className="d-flex align-center">
            creating
            <input
              onChange={this.props.setServings}
              value={this.props.servings}
              className="my-input ml-2"
              type="text"
              placeholder="4 to 6 servings"
              name="servings"
            />
          </label>
          <label className="d-flex align-center">
            in
            <input
              onChange={this.props.setTime}
              value={this.props.time}
              className="my-input ml-2"
              type="text"
              placeholder="1 hour"
              name="time"
            />
          </label>
        </div>

        <section className="ingredients-preparation-grid">
          <div>
            <h2 className="title is-3 mb-1 font-family-title bold">
              Ingredients
            </h2>
            <ul>
              {this.props.ingredients.map((x, i) => (
                <Ingredient
                  key={x.name + i}
                  recipeID={-1}
                  id={i}
                  update={(ingre: IIngredientBasic) =>
                    this.props.updateIngredient(i, ingre)
                  }
                  remove={() => this.props.removeIngredient(i)}
                  quantity={x.quantity}
                  optional={x.optional}
                  name={x.name}
                  description={x.description}
                />
              ))}
            </ul>

            <AddIngredientForm
              handleAddIngredient={addIngredient}
              cancelAddIngredient={cancelAddIngredient}
              handleInputChange={handleIngredientChange}
              quantity={quantity}
              name={name}
              description={description}
              optional={optional}
              error={errorWithIngredients}
            />
          </div>

          <div>
            <h2 className="title is-3 mb-1 font-family-title bold">
              Preparation
            </h2>
            <ul>
              {this.props.steps.map((s, i) => (
                <div key={s.text + i}>
                  <label className="better-label">Step {i + 1}</label>
                  <ListItem
                    id={i}
                    text={s.text}
                    update={this.props.updateStep}
                    delete={this.props.removeStep}
                  />
                </div>
              ))}
            </ul>
            <AddStepForm
              handleInputChange={handleInputChange}
              addStep={addStep}
              cancelAddStep={cancelAddStep}
              stepNumber={this.props.steps.length + 1}
              text={step}
              error={errorWithSteps}
            />
          </div>
        </section>
        <div className="d-flex justify-space-between align-items-center">
          <button
            className="my-button"
            onClick={this.props.clearForm}
            name="create recipe">
            Clear
          </button>

          <div className="d-flex justify-space-between">
            <label className="d-flex align-center">
              for
              <div className="select ml-2 is-small">
                <select
                  disabled={this.props.loadingTeams}
                  value={this.props.teamID || "personal"}
                  onChange={this.handleTeamChange}>
                  <option value="personal">Personal</option>
                  {this.props.teams.map(t => (
                    <option key={t.id} value={t.id}>
                      Team: {t.name}
                    </option>
                  ))}
                </select>
              </div>
            </label>

            <ButtonPrimary
              className="ml-2"
              type="submit"
              onClick={handleSubmit}
              name="create recipe"
              loading={this.props.loading}>
              Create Recipe
            </ButtonPrimary>
          </div>
        </div>
      </div>
    )
  }
}
