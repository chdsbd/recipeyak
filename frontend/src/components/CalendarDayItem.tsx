import React from "react"
import { styled } from "@/theme"
import { Link } from "react-router-dom"
import { useDrag } from "react-dnd"
import { isInsideChangeWindow } from "@/date"
import { recipeURL } from "@/urls"
import { DragDrop } from "@/dragDrop"
import { IRecipe } from "@/store/reducers/recipes"
import { ICalRecipe } from "@/store/reducers/calendar"
import { TextInput } from "@/components/Forms"
import { Result, isOk } from "@/result"
import { useGlobalEvent } from "@/hooks"
import { CalendarDayItemModal } from "@/components/CalendarDayItemModal"

const COUNT_THRESHOLD = 1

interface ICountProps {
  readonly value: number
  readonly onChange: (e: React.ChangeEvent<HTMLInputElement>) => void
}
function Count({ value: count, onChange }: ICountProps) {
  if (count > COUNT_THRESHOLD) {
    return (
      <div className="d-flex">
        <TextInput
          className="text-small text-right w-2rem"
          name="calendar-item-count"
          onChange={onChange}
          value={count}
        />
      </div>
    )
  }
  return null
}

interface IRecipeLink {
  readonly id: IRecipe["id"] | string
  readonly name: IRecipe["name"]
  readonly onClick: (e: React.MouseEvent) => void
}

const StyledLink = styled(Link)`
  line-height: 1.3;
  font-size: ${props => props.theme.text.small};
  word-break: break-word;
  background-color: ${props => props.theme.color.background};
  border-radius: 5px;
  padding: 0.35rem;
  font-weight: 600;
`

function RecipeLink({ name, id, onClick }: IRecipeLink) {
  const to = recipeURL(id, name)
  return (
    <StyledLink to={to} onClick={onClick}>
      {name}
    </StyledLink>
  )
}

interface ICalendarListItemProps {
  readonly visibility: React.CSSProperties["visibility"]
}

const CalendarListItem = styled.li<ICalendarListItemProps>`
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 0.5rem;
  visibility: ${props => props.visibility};
`

export interface ICalendarItemProps {
  readonly count: ICalRecipe["count"]
  readonly remove: () => void
  readonly updateCount: (
    count: ICalRecipe["count"],
  ) => Promise<Result<void, void>>
  readonly refetchShoppingList: () => void
  readonly date: Date
  readonly recipeID: IRecipe["id"] | string
  readonly recipeName: IRecipe["name"]
  readonly scheduledId: ICalRecipe["id"]
  readonly teamID: TeamID
}

export function CalendarItem({
  count: propsCount,
  updateCount: propsUpdateCount,
  date,
  refetchShoppingList,
  remove,
  recipeName,
  recipeID,
  teamID,
  scheduledId,
}: ICalendarItemProps) {
  const [count, setCount] = React.useState(propsCount)
  const ref = React.useRef<HTMLLIElement>(null)
  const [show, setShow] = React.useState(false)

  React.useEffect(() => {
    setCount(propsCount)
  }, [propsCount])

  const updateCount = (newCount: number) => {
    if (!isInsideChangeWindow(date)) {
      return
    }
    const oldCount = count
    setCount(newCount)
    propsUpdateCount(newCount).then(res => {
      if (isOk(res)) {
        refetchShoppingList()
      } else {
        setCount(oldCount)
      }
    })
  }

  const handleKeyPress = (e: KeyboardEvent) => {
    if (!ref.current?.matches(":hover")) {
      return
    }

    if (!isInsideChangeWindow(date)) {
      return
    }

    if (e.key === "#" || e.key === "Delete") {
      remove()
    }
    if (e.key === "A" || e.key === "+") {
      updateCount(count + 1)
    }
    if (e.key === "X" || e.key === "-") {
      updateCount(count - 1)
    }
  }

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    updateCount(parseInt(e.target.value, 10))

  const dragItem: ICalendarDragItem = {
    type: DragDrop.CAL_RECIPE,
    recipeID,
    count,
    scheduledId,
    date,
  }

  const [{ isDragging }, drag] = useDrag({
    item: dragItem,
    end: (_dropResult, monitor) => {
      // when dragged onto something that isn't a target, we remove it
      // but we don't remove when in past as we only copy from the past
      if (!monitor.didDrop() && isInsideChangeWindow(date)) {
        remove()
      }
    },
    collect: monitor => {
      return {
        isDragging: monitor.isDragging(),
      }
    },
  })

  const visibility =
    isDragging && isInsideChangeWindow(date) ? "hidden" : "visible"

  useGlobalEvent({ keyUp: handleKeyPress })

  drag(ref)

  return (
    <>
      <CalendarListItem ref={ref} visibility={visibility}>
        <RecipeLink
          name={recipeName}
          id={recipeID}
          onClick={e => {
            e.preventDefault()
            setShow(true)
          }}
        />
        <Count value={count} onChange={handleChange} />
      </CalendarListItem>
      {show ? (
        <CalendarDayItemModal
          scheduledId={scheduledId}
          teamID={teamID}
          recipeName={recipeName}
          recipeId={recipeID}
          date={date}
          onClose={() => setShow(false)}
        />
      ) : null}
    </>
  )
}

export interface ICalendarDragItem
  extends Pick<ICalendarItemProps, "recipeID" | "count" | "scheduledId" | "date"> {
  readonly type: DragDrop.CAL_RECIPE
}
