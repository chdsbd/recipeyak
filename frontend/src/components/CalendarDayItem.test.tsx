import React from "react"
import renderer from "react-test-renderer"
import { CalendarItem } from "@/components/CalendarDayItem"
import HTML5Backend from "react-dnd-html5-backend"
import { DndProvider } from "react-dnd"
import { MemoryRouter } from "react-router"
import { ThemeProvider, theme } from "@/theme"
import { isPast } from "date-fns"

const Provider: React.FC = ({ children }) => (
  <ThemeProvider theme={theme}>
    <MemoryRouter>
      <DndProvider backend={HTML5Backend}>{children}</DndProvider>
    </MemoryRouter>
  </ThemeProvider>
)

describe("<CalendarDayItem> Snap", () => {
  test("smoke test render with styled components", () => {
    const scheduledRecipe = {
      id: 25,
      recipe: {
        id: 10,
        name: "Baked Ziti"
      }
    }

    const pastDate = new Date(1776, 1, 1)
    expect(isPast(pastDate)).toBe(true)

    const tree = renderer
      .create(
        <Provider>
          <CalendarItem
            count={1}
            id={scheduledRecipe.id}
            recipeID={scheduledRecipe.recipe.id}
            recipeName={scheduledRecipe.recipe.name}
            remove={jest.fn()}
            updateCount={jest.fn()}
            refetchShoppingList={jest.fn()}
            date={pastDate}
          />
        </Provider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
