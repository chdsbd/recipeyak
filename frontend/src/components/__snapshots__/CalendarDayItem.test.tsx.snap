// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalendarDayItem> Snap smoke test render with styled components 1`] = `
.c0 {
  line-height: 1.3;
  font-size: 0.875rem;
  word-break: break-word;
  background-color: #ff7247;
  border-radius: 5px;
  padding: 0.35rem;
  color: whitesmoke;
  font-weight: 600;
}

.c0:hover {
  color: whitesmoke;
}

<li
  className="d-flex align-items-center cursor-pointer justify-space-between mb-2"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "visibility": "visible",
    }
  }
>
  <a
    className="c0"
    href="/recipes/10-baked-ziti"
    onClick={[Function]}
  >
    Baked Ziti
  </a>
</li>
`;
